sdk/tests/conformance/buffers/buffer-data-and-buffer-sub-data.html

```diff
@@ -0,0 +1,163 @@
+** IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+** CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
+** TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
+** MATERIALS OR THE USE OR OTHER DEALINGS IN THE MATERIALS.
+*/
+
+-->
+
+<!DOCTYPE html>
+<html>
+<head>
+<meta charset="utf-8">
+<link rel="stylesheet" href="../../resources/js-test-style.css"/>
+<script src="../../js/js-test-pre.js"></script>
+<script src="../../js/webgl-test-utils.js"></script>
+</head>
+<body>
+<div id="description"></div>
+<div id="console"></div>
+
+<script>
+"use strict";
+description("Test bufferData/bufferSubData with or without ArrayBuffer input");
+
+debug('Regression test for <a href="https://bugs.webkit.org/show_bug.cgi?id=41884">https://bugs.webkit.org/show_bug.cgi?id=41884</a> : <code>Implement bufferData and bufferSubData with ArrayBuffer as input</code>');
+
+var wtu = WebGLTestUtils;
+var gl = wtu.create3DContext();
+var buf = null;
+var array = null;
+var arrayZero = null;
+if (!gl) {
+    testFailed("WebGL context does not exist");
+} else {
+    testPassed("WebGL context exists");
+
+    bufferDataTest();
+    bufferDataTestWithArrayBuffer();
+
+    bufferSubDataTest();
+}
+
+function bufferDataTest() {
+    debug("");
+    debug("Test bufferData without ArrayBuffer input");
+
+    buf = gl.createBuffer();
+    shouldBeNonNull("buf");
+
+    gl.bufferData(gl.ARRAY_BUFFER, 4, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION);
+
+    gl.bindBuffer(gl.ARRAY_BUFFER, buf);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.bufferData(gl.ARRAY_BUFFER, -4, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE);
+
+    gl.bufferData(gl.ARRAY_BUFFER, 0, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.bufferData(gl.ARRAY_BUFFER, 4, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.bindBuffer(gl.ARRAY_BUFFER, null);
+}
+
+function bufferDataTestWithArrayBuffer() {
+    debug("");
+    debug("Test bufferData with ArrayBuffer input");
+
+    array = new ArrayBuffer(128);
+    shouldBeNonNull("array");
+
+    buf = gl.createBuffer();
+    shouldBeNonNull("buf");
+
+    gl.bufferData(gl.ARRAY_BUFFER, array, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.INVALID_OPERATION);
+
+    gl.bindBuffer(gl.ARRAY_BUFFER, buf);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    gl.bufferData(gl.ARRAY_BUFFER, null, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.INVALID_VALUE);
+
+    // This should not crash, but the selection of the overload is ambiguous per Web IDL.
+    gl.bufferData(gl.ARRAY_BUFFER, null, gl.STATIC_DRAW);
+    gl.getError();
+
+    gl.bufferData(gl.ARRAY_BUFFER, array, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
+
+    arrayZero = new ArrayBuffer(0);
+    shouldBeNonNull("arrayZero");
+    gl.bufferData(gl.ARRAY_BUFFER, arrayZero, gl.STATIC_DRAW);
+    wtu.glErrorShouldBe(gl, gl.NO_ERROR);
```

Comment:
What do the other WebGL implementations (Firefox's, Apple's, Microsoft's) do in this case? Since passing a null ArrayBuffer is disallowed, and Chrome's implementation has historically not allowed 0-sized buffers to be allocated, I think this should generate an INVALID_VALUE error, and be specified to do so.